#########################################
# Figure 1E,F,G plot
# This script examines the whole-genome methylation profile between SCLC cfDNA and NCC (non-cancer control) cfDNA
#
# Written by Sami Ul Haq (Dec 2021)

# The PCA was generated by examining 300bp ENCODE-blacklisted filtered, whole-genome methylation profiles between non-cancer controls (NCCs) and SCLC patient cfDNA using DESeq2 on an HPC node with 180GB memory. The DESeq2 results object was saved and is examined in this script.
# See Zenodo link for downloading data:
# Ul Haq, Sami, Tsao, Ming S., Cabanero, Michael, de Carvalho, Daniel, Liu, Geoffrey, Bratman, Scott V., & Lok, Benjamin H. (2022). Processed counts data of cfMeDIP-seq profiles of small cell lung cancer patients [Data set]. In Cell Press iScience (Vol. 25, Number 12). Zenodo. https://doi.org/10.5281/zenodo.7235989

# necessary libraries
source("../data/GSEA_KEGG_Analysis_Functions.R")
library(ggplot2)
library(reshape2)
library(Repitools)
library(BSgenome.Hsapiens.UCSC.hg19)

# The DESEQ2 analysis examines SCLC vs NCCs. This means log2(NCCs / SCLC)
# So, (+)Log2FC (and p-val < 0.05) = windows enriched in NCCs cfDNA
# and (-)Log2FC (and p-val < 0.05) = windows enriched in SCLC cfDNA

load("../data/DMR.analysis.NCC.cfDNA.vs.SCLC.cfDNA.RData")

# cutoffs for subsequent analyses
significance.cutoff <- 0.05
fc.cutoff <- 1

dsq2 <- data.frame(log2FC=sami.healthy.vs.SCLC.cfDNA$log2FoldChange, adjust.pval=sami.healthy.vs.SCLC.cfDNA$padj)
dsq2 <- na.omit(dsq2)

dsq2$log2FC <- -( dsq2$log2FC )

dsq2$col <- rep("NS", length(dsq2$log2FC))
dsq2$col[which(dsq2$adjust.pval < 0.05 & dsq2$log2FC > 1)] <- "SCLC"
dsq2$col[which(dsq2$adjust.pval < 0.05 & dsq2$log2FC < -1)] <- "Non-cancer"

named.color.vector <- c("NS" = "#BAC4C8", "SCLC" = "#E21F26", "Non-cancer" = "#4EAF49")

vulcan <- ggplot(dsq2, aes(x=log2FC, y=-log10(adjust.pval), color=col)) +
  geom_point(alpha=0.5, size=5) +
  scale_color_manual(values = named.color.vector) +
  theme_bw() +
  #ggtitle("Examining Significant DMRs in SCLC All vs Healthy Donors") +
  geom_vline(xintercept=1, linetype=1, size=2, color="#B79896") +
  geom_vline(xintercept=-1, linetype=1, size=2, color="#B79896") +
  geom_hline(yintercept = -log10(0.05), linetype=1, size=2, color="#B79896") +
  theme(axis.text.x = element_text(size = 24),
        axis.text.y = element_text(size = 24),
        axis.title.y = element_text(size = 26),
        axis.title.x = element_text(size = 26),
        legend.position = "right") +
  coord_cartesian(xlim = c(-2.5, 6))

ggsave(plot=vulcan,
       filename = "../results/Figure_1E.tiff",
       units = "in",
       width = 6,
       height = 7,
       dpi = 300)

# the plot is deleted after saving because it is memory intensive
rm(vulcan, dsq2)
gc()

#################################################################################
# The following examines the characteristics of the windows that are hypermethylated in SCLC patients vs Healthy Donors. Specifically, to get an idea of genomic locations commonly methylated (i.e. promoter regions, gene bodies, etc.)

regions.in.SCLC.cfDNA <- rownames(sami.healthy.vs.SCLC.cfDNA)[which(sami.healthy.vs.SCLC.cfDNA$padj<significance.cutoff & sami.healthy.vs.SCLC.cfDNA$log2FoldChange < -fc.cutoff)]
regions.in.SCLC.cfDNA.grange <- medip.regions.to.GRange.maker(regions.in.SCLC.cfDNA)

regions.in.Healthy <- rownames(sami.healthy.vs.SCLC.cfDNA)[which(sami.healthy.vs.SCLC.cfDNA$padj<significance.cutoff & sami.healthy.vs.SCLC.cfDNA$log2FoldChange > fc.cutoff)]
regions.in.Healthy.grange <- medip.regions.to.GRange.maker(regions.in.Healthy)

# annotation hub annotations built
annots = c("hg19_cpgs")
annotations = build_annotations(genome = 'hg19', annotations = annots)

All.SCLC.regions.annotated <- genomic.window.annotator(queryGrange = regions.in.SCLC.cfDNA.grange, subjectGrange = annotations, removeDuplicateRegions = FALSE)
# cpg features in SCLC are saved
table.of.sclc.cpg.feat <- c(table(All.SCLC.regions.annotated$type))

Healthy.regions.annotated <- genomic.window.annotator(queryGrange = regions.in.Healthy.grange, subjectGrange = annotations, removeDuplicateRegions = FALSE)
# cpg features in healthy are saved
table.of.healthy.cpg.feat <- c(table(Healthy.regions.annotated$type))

# the cpg features plot is made below

features.SCLC <- table.of.sclc.cpg.feat / sum(table.of.sclc.cpg.feat)
features.Healthy <- table.of.healthy.cpg.feat / sum(table.of.healthy.cpg.feat)


# empty dataframe is made

temp.annots.holder <- data.frame(row.names = 1:length(features.SCLC), stringsAsFactors = FALSE)
temp.annots.holder$Features <- as.vector( names(features.SCLC))
temp.annots.holder$SCLC <- rep(0, length(features.SCLC) )
temp.annots.holder$HEALTHY <- rep(0, length(features.SCLC) )

#dataframe is populated with the appropriate frequencies of each annotation 
temp.annots.holder$SCLC[match( names(features.SCLC),  temp.annots.holder$Features)] <- features.SCLC
temp.annots.holder$HEALTHY[match( names(features.Healthy),  temp.annots.holder$Features)] <- features.Healthy

colnames(temp.annots.holder) <- c("Features", "SCLC.cfDNA", "Healthy.cfDNA")

reform.annots.df <- melt(temp.annots.holder, measure.vars = c("SCLC.cfDNA", "Healthy.cfDNA"), variable.name = "Groups")

cpg.graph <- ggplot(reform.annots.df, aes(Groups, value, fill=as.factor(Features))) + 
  geom_col(width = 0.8) +
  theme_classic() +
  labs(fill="CpG Feature") +
  theme(axis.text.x = element_text()) +
  scale_fill_manual(values = c("hg19_cpg_inter"="#66c2a5", 
                               "hg19_cpg_islands"="#fc8d62", 
                               "hg19_cpg_shelves"="#8da0cb", 
                               "hg19_cpg_shores"="#e78ac3")) +
  scale_x_discrete(labels=c("SCLC cfDNA", "Non-cancer donor cfDNA")) +
  ylab("Proportion of Significant DMRs") +
  theme(legend.position = "right",
        axis.text.x = element_text(size=18, angle = 0),
        axis.text.y = element_text(size = 18),
        axis.title.y = element_text(size = 20),
        axis.title.x = element_blank(),
        title = element_text(size = 24),
        legend.text = element_text(size=18)) +
  scale_y_continuous(labels = scales::percent)

# pdf
ggsave(plot=cpg.graph, filename="../results/Figure_1F.pdf", width=8, height=8, units = "in")

#################################################################################
# Permutation analysis is done here
# Basically, how often are features, i.e. CpG Islands (a feature of interest), picked up in SCLC cfDNA data compared to the genome in general. To do this, first need to examine the basal (null distribution) occurence of CpG features in the human genome. For this, 60000 windows (this was the number of significant DMRs observed in SCLC cfDNA compared to NCC cfDNA) are sampled randomly from the entire genome (1000 times).

annots = c("hg19_cpgs")
annotations = build_annotations(genome = 'hg19', annotations = annots)

annotations.cpg.islands <- subset(annotations, annotations$type %in% "hg19_cpg_islands")
annotations.cpg.open.sea <- subset(annotations, annotations$type %in% "hg19_cpg_inter")
annotations.cpg.shelves <- subset(annotations, annotations$type %in% "hg19_cpg_shelves")
annotations.cpg.shores <- subset(annotations, annotations$type %in% "hg19_cpg_shores")

# makes binned 300bp genome
binned.genome <- genomeBlocks(BSgenome.Hsapiens.UCSC.hg19, chrs=seqnames(BSgenome.Hsapiens.UCSC.hg19)[1:22], width=300)

# holds the cpg islands occurrence in the null distribution (overall human genome, not just cfMeDIP data)
cpg.islands.in.null <- c()

# holds the open seas occurrence in the null distribution (overall human genome, not just cfMeDIP data)
open.seas.in.null <- c()

# holds the cpg shelves occurrence in the null distribution (overall human genome, not just cfMeDIP data)
cpg.shelves.in.null <- c()

# holds the cpg shores occurrence in the null distribution (overall human genome, not just cfMeDIP data)
cpg.shores.in.null <- c()


# number of times to sample the human genome
num.of.iterations <- 1000

for(iteration in 1:num.of.iterations) {
  
  # sampling this many windows at a time because ~60000 windows were upregulated in SCLC compared to NCCs
  sample.number <- length(which(sami.healthy.vs.SCLC.cfDNA$padj<0.05 & sami.healthy.vs.SCLC.cfDNA$log2FoldChange < -1))
  
  sample.genome <- sample(binned.genome, sample.number, replace=FALSE)
  
  # calculates the CpG Islands occurrence
  num.of.cpgs.islands <- findOverlapPairs(annotations.cpg.islands, sample.genome)
  proportions.of.islands <- length(num.of.cpgs.islands)/sample.number
  cpg.islands.in.null <- c(cpg.islands.in.null, proportions.of.islands)
  
  # calculates the open sea regions occurrence
  num.of.open.seas <- findOverlapPairs(annotations.cpg.open.sea, sample.genome)
  proportions.of.open.seas <- length(num.of.open.seas)/sample.number
  open.seas.in.null <- c(open.seas.in.null, proportions.of.open.seas)
  
  # calculates the CpG shelves regions occurrence
  num.of.cpg.shelves <- findOverlapPairs(annotations.cpg.shelves, sample.genome)
  proportions.of.shelves <- length(num.of.cpg.shelves)/sample.number
  cpg.shelves.in.null <- c(cpg.shelves.in.null, proportions.of.shelves)
  
  # calculates the CpG shores regions occurrence
  num.of.cpg.shores <- findOverlapPairs(annotations.cpg.shores, sample.genome)
  proportions.of.shores <- length(num.of.cpg.shores)/sample.number
  cpg.shores.in.null <- c(cpg.shores.in.null, proportions.of.shores)
}

#reform.annots.df

# converts the occurence of CpG Islands in my SCLC data to a Z score
cpg.island.occurence.in.SCLC <- ( reform.annots.df$value[2] - mean(cpg.islands.in.null) ) / sd(cpg.islands.in.null)
# converts the occurence of open seas in my SCLC data to a Z score
open.sea.occurence.in.SCLC <- (reform.annots.df$value[1] - mean(open.seas.in.null) ) / sd(open.seas.in.null)
cpg.shelves.occurence.in.SCLC <- (reform.annots.df$value[3] - mean(cpg.shelves.in.null)) / sd(cpg.shelves.in.null)
cpg.shores.occurence.in.SCLC <- (reform.annots.df$value[4] - mean(cpg.shores.in.null)) / sd(cpg.shores.in.null)


# scale() converts the raw scores to a Z score (normalizes)
cpg.islands.in.null.z <- scale(cpg.islands.in.null)
open.seas.in.null.z <- scale(open.seas.in.null)
cpg.shelves.in.null.z <- scale(cpg.shelves.in.null)
cpg.shores.in.null.z <- scale(cpg.shores.in.null)


# holds the permutations for the null distributions
permutations.df <- data.frame(null.cpg.islands = cpg.islands.in.null.z,
                              null.cpg.shores = cpg.shores.in.null.z,
                              null.cpg.shelves = cpg.shelves.in.null.z,
                              null.open.seas = open.seas.in.null.z)

# reshapes df to make it ggplot friendly
permutations.df <- melt(permutations.df,
                        measure.vars = c("null.cpg.islands", "null.cpg.shores", "null.cpg.shelves", "null.open.seas"), 
                        variable.name = "CpG.Feature")

# the sclc specific z-scores are saved in df to overlay
sclc.scores.df <- data.frame(CpG.Feature = c("null.cpg.islands", 
                                             "null.cpg.shores", 
                                             "null.cpg.shelves", 
                                             "null.open.seas"),
                             CpG.Feature.Score =c(cpg.island.occurence.in.SCLC,
                                                  cpg.shores.occurence.in.SCLC,
                                                  cpg.shelves.occurence.in.SCLC,
                                                  open.sea.occurence.in.SCLC))

permutations.sclc.plot <- ggplot(data=permutations.df, aes(x=CpG.Feature, y=value)) +
                            geom_boxplot() +
                            geom_point(data=sclc.scores.df, aes(x=CpG.Feature, CpG.Feature.Score), col="red", shape=18, size=6) +
                            theme_bw() +
                            theme(legend.position = "right",
                                  axis.text.x = element_text(size = 18),
                                  axis.text.y = element_text(size = 18),
                                  axis.title.y = element_text(size = 20),
                                  axis.title.x = element_blank(),
                                  title = element_text(size = 24),
                                  legend.text = element_text(size=18)) +
                            scale_x_discrete(labels=c("Islands", "Shores", "Shelves", "Open Sea")) +
                            ylab("Z-score")

ggsave(plot=permutations.sclc.plot, filename="../results/Figure_1G.pdf", width=6, height=7, units = "in")
